####-rfc-generalized-unroller-and-equivalence-library.md

```diff
@@ -0,0 +1,322 @@
+
+Two `EquivalenceLibrary` instances will be automatically instantiated during each Qiskit session:
+
+-   the `StandardEquivalenceLibrary` will be pre-constructed for the gates in the standard library,
+-   and a `SessionEquivalenceLibrary` will be constructed and will be the source and target of definitions for user-defined gates.
+
+`EquivalenceLibrary`s can be singly nested, so derived libraries layer modifications on top of an existing library without affecting the original.
+
+The set of circuits returned in an equivalence entry by the library does not have a defined order, the library consumer should inspect the set to find a desired circuit,.
+
+If an input gate is parameterized by a set of `qiskit.circuit.Parameters`, the returned circuit will be parameterized by the same `Parameter` set.
+
+A consequence of this approach is that setting custom definitions on individual `Instruction` instances will no longer be supported.
+In the future, this can be replaced by incorporating `EquivalenceLibrary` entries to be tied to an `Instruction` label or instance.
+
+
+## Example: Fetching standard definition
+
+    from qiskit.extensions.standard import CzGate, CnotGate, ToffoliGate
+    from qiskit.extensions.standard import StandardEquivalenceLibrary as sel
+    
+    cz_entry = sel.get_entry(CzGate())
+    cz_entry[0].draw()
+    
+    sel.get_entry(ToffoliGate())[0].draw(output='mpl')
+
+## `qiskit.transpiler.passes.BasisTranslator`
+
+As a replacement for the current `Unroller` pass, we introduce a `BasisTranslator` pass which utilizes a provided `EquivalenceLibrary` to find a set of translations to map from a source basis to a target basis via a bramble-search algorithm, as compared to a tree-descent search in the existing `Unroller`.
+
+# Alternative Approaches
+
+# Unresolved questions
+
+Gates which are defined over a variable number of qubits, such as the Mølmer–Sørensen gate, may require additional handling in the `EquivalenceLibrary` interface.
+Entries could be defined as callables rather than circuits, either generally or specifically for variadic gates, though this may hinder the ability of transpiler passes to inspect available decompositions and intelligently choose among them.
+Storing one circuit entry per gate width in the library is another option, growing the library size proportional to the number of variadic gates and their maximum widths.
+
+The `Decompose` transpile pass and corresponding `QuantumCircuit.decompose` method depend on a single unique path through `Instruction.definition`.
+These methods could be deprecated and removed, maintained via a decomposition search targeting `U3` and `CX`, or maintained via additional library entry labeling to support identification of one entry as the legacy `definition`.
```

Comment:
The use case for `decompose()` is when one builds a circuit using multiple composite gates (just because it's easier to reuse a composite gate, not because one is interested in different decompositions of that composite gate). If we don't have `decompose` then you have to specify a basis to unroll to. How would I select the basis to yield the current `decompose` behavior? If we can solve this then I'm fine deprecating decompose (i.e. a canonical definition for gates), and going with this more general framework.

Reply:
Adding an optional target basis kwarg to `decompose` defaulting to `['u3', 'cx']` would preserve the current behavior. Maybe better for the use case you're describing though would be to replace the current behavior with something like 'decompose only gates from outside the standard library'.